runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java

```diff
@@ -76,6 +76,7 @@
   private static final Logger LOG = LoggerFactory.getLogger(UnboundedReadFromBoundedSource.class);
 
   private final BoundedSource<T> source;
+  private Coder<T> defaultOutputCoder;
```

Comment:
Transform classes can not have mutable state. Moreover, having getDefaultOutputCoder() return different values before and after expand() was called seems very brittle - the SDK or runners make no guarantees about the order in which methods will be called on a PTransform (instead, we require that PTransform's are immutable, so the concept of call order is irrelevant)